### PR TITLE
media-gfx/blender Fix live ebuild and add support for nanovdb.

### PR DESCRIPTION
**Pull request by DarkDefender**
@waebbl Do you have time to look into CUDA support for nanovdb?
I don't have and CUDA card to test and implement it myself.

I talked to Brecht (the main Cycles dev), the told me the only way to actually check if nanovdb support is working is to render a scene with volumes. If it works, the smoke should render, otherwise it should just be empty/corrupted.

You could be able to test this with the `tests/render/openvdb/openvdb_smoke.blend` in the automated test bundle.
The reason Blender uses nanovdb for rendering in cycles is that it should reduce memory usage by a bit.
So that is something you could look out for too.

### CODE REVIEW

File: media-gfx/openvdb/openvdb-9.0.0-r2.ebuild [hunk 1]
```diff
@@ -14,23 +14,23 @@ SRC_URI="https://github.com/AcademySoftwareFoundation/${PN}/archive/v${PV}.tar.g
...
-	^^ ( abi6-compat abi7-compat abi8-compat )
+	^^ ( abi6-compat abi7-compat abi8-compat abi9-compat )
 	python? ( ${PYTHON_REQUIRED_USE} )
 "
 RDEPEND="
 	>=dev-cpp/tbb-2021.4.0:=
 	dev-libs/boost:=
 	dev-libs/jemalloc:=
 	dev-libs/log4cplus:=
-	>=dev-libs/imath-3.1.4-r2:=
```
Comment (thesamesam) [inline media-gfx/openvdb/openvdb-9.0.0-r2.ebuild:30]: Keep this to ensure we're using an imath without the bad headers (Imath-3 vs Imath)

**Review by thesamesam — COMMENTED** [review]
Nice to see you back! I hope things are better now!

File: media-gfx/openvdb/openvdb-9.0.0-r2.ebuild (hunk 1 above)
Comment (DarkDefender) [reply media-gfx/openvdb/openvdb-9.0.0-r2.ebuild:30]: I must have changed and included these two by mistake. I will revert these two changes.

**Comment by DarkDefender** [comment]
> Nice to see you back! I hope things are better now!

It is starting to get a bit better. But I think I might be a bit flaky for a quite some time still.
But I'll try to get in some overdue changes. 

Thanks again for taking care of so much of the maintenance!

File: media-gfx/blender/blender-9999.ebuild [hunk 2]
```diff
@@ -90,7 +90,7 @@ RDEPEND="${PYTHON_DEPS}
 	)
 	opensubdiv? ( >=media-libs/opensubdiv-3.4.0[cuda=] )
 	openvdb? (
-		>=media-gfx/openvdb-8.2.0-r2:=
+		>=media-gfx/openvdb-8.2.0-r2:=[nanovdb?]
```
Comment (waebbl) [inline media-gfx/blender/blender-9999.ebuild:93]: Seeing that you only added the nanovdb USE flag to v9.0.0, shouldn't we restrict this to >=openvdb-9.x?

**Comment by waebbl** [comment]
@DarkDefender I can test the nanovdb feature and if it's working with cuda.

**Comment by DarkDefender** [comment]
> @DarkDefender I can test the nanovdb feature and if it's working with cuda.

Thanks!
Note that if we add CUDA support, we should probably add a `cuda` useflag as well as they have a switch for this in nanovdb:
https://academysoftwarefoundation.github.io/openvdb/NanoVDB_HowToBuild.html

Though it should now autodetect cuda if you have the libraries installed.
So you should be able to test without editing the ebuild.

**Comment by waebbl** [comment]
> > @DarkDefender I can test the nanovdb feature and if it's working with cuda.
> 
> Thanks! Note that if we add CUDA support, we should probably add a `cuda` useflag as well as they have a switch for this in nanovdb: https://academysoftwarefoundation.github.io/openvdb/NanoVDB_HowToBuild.html
> 
> Though it should now autodetect cuda if you have the libraries installed. So you should be able to test without editing the ebuild.

Thanks for the hint. I'll take a look at the build system and see about cuda support. Noticed the nanovdb option already during a earlier patch of the package, but lacked the time to implement it.

File: media-gfx/blender/blender-9999.ebuild [hunk 3]
```diff
@@ -263,6 +257,7 @@ src_configure() {
 		-DWITH_SDL=$(usex sdl)
 		-DWITH_STATIC_LIBS=OFF
 		-DWITH_SYSTEM_EIGEN3=ON
+		-DWITH_SYSTEM_FREETYPE=ON
```
[the remark this answers is not in the document]
Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild:260]: The variable doesn't say what it does -_-

There is no bundled freetype in blender.
The flag currently only makes the `brotli` library optional.
I have poked the upstream developer who introduced this change to clarify and hopefully change this.

File: media-gfx/blender/blender-9999.ebuild (hunk 2 above)
Comment (DarkDefender) [reply media-gfx/blender/blender-9999.ebuild:93]: Yeah, I think I'll just be lazy with this and bump support to 9.x as you suggested.
Cleaner that way.

**Comment by waebbl** [comment]
This patch to your PR builds fine and seems to work. At least the image looks identical to one created with openvdb-8 / blender-3.0 and the smoke is visible, so I assume the cuda functions from nanovdb are working.

[0001-media-gfx-openvdb-enable-CUDA-support-for-NanoVDB.patch.txt](https://github.com/gentoo/gentoo/files/8098770/0001-media-gfx-openvdb-enable-CUDA-support-for-NanoVDB.patch.txt)

**Comment by DarkDefender** [comment]
> so I assume the cuda functions from nanovdb are working.
I assume you rendered the cycles scene with CUDA and not the CPU render?
Just double checking, I've made that mistake in the past when looking for GPU render issues.

Thanks for the patch! I've integrated it into this PR.

**Comment by waebbl** [comment]
> > so I assume the cuda functions from nanovdb are working.
> 
> I assume you rendered the cycles scene with CUDA and not the CPU render? Just double checking, I've made that mistake in the past when looking for GPU render issues.
> 
> Thanks for the patch! I've integrated it into this PR.

Yes I did :) At first CPU was active, but then I wondered about the low memory footprint and checked the preferences.

BTW, a comparison on memory usage leads to only a minor reduction with nanovdb for this scene. openvdb-9 with nanovdb needs ~360M and openvdb-8 without nanovdb needs ~370M.

File: media-gfx/openvdb/openvdb-9.0.0-r3.ebuild [hunk 4]
```diff
@@ -14,16 +14,18 @@ SRC_URI="https://github.com/AcademySoftwareFoundation/${PN}/archive/v${PV}.tar.g
...
-	numpy? ( python )
 	^^ ( abi6-compat abi7-compat abi8-compat abi9-compat )
+	cuda? ( nanovdb )
+	numpy? ( python )
 	python? ( ${PYTHON_REQUIRED_USE} )
 "
+
 RDEPEND="
-	>=dev-cpp/tbb-2021.4.0:=
+	>=dev-cpp/tbb-2020.3
```
Comment (thesamesam) [inline media-gfx/openvdb/openvdb-9.0.0-r3.ebuild:28]: dropped :=

Comment (waebbl) [reply media-gfx/openvdb/openvdb-9.0.0-r3.ebuild:28]: 2020.3 doesn't have subslots

Comment (thesamesam) [reply media-gfx/openvdb/openvdb-9.0.0-r3.ebuild:28]: It has an implicit 0/0 which then, when upgrading to a newer version, is a change (e.g. 0/2021 or whatever), so it matters if we know a newer version has them, as we want to get rebuilt when the ABI changes

Comment (waebbl) [reply media-gfx/openvdb/openvdb-9.0.0-r3.ebuild:28]: i see

**Pull request closed without merging**